torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,116 @@
+import itertools
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class LazyModuleMixin:
+    """A mixin for modules that lazily initialize buffers and parameters.
```

Comment:
The location for this docstring might not be great as users won't check this doc in general but only the `LazyLinear` one (or corresponding Module they use).

If we only do `LazyLinear` for now, we should just move this to that docstring.
When we add convolution later, maybe this shared part can be moved to a note in the doc that each lazy module we implement refers to.


Also an important part of this doc should be when we need to do a dummy forward pass or not.
In particular it is not required before creating optimizers or serializing but it is unclear what should be done when the user wants to do custom initialization of their weights, or use DataParallel, or DistributedDataParallel, or reparametrization like `nn.utils.weight_norm()`.

Reply:
Agreed, having the doc on when we need to do a dummy forward pass is super important for setting expectations :)